mocks/firestore.js

```diff
@@ -116,6 +119,60 @@ class FakeFirestore {
     mockRunTransaction(...arguments);
     return updateFunction(new FakeFirestore.Transaction());
   }
+
+  _updateData(path, object, merge) {
+    // Do not update unless explicity set to mutable.
+    if (!this.options.mutable) {
+      return;
+    }
+
+    // note: this logic could be deduplicated
+    const pathArray = path
+      .replace(/^\/+/, '')
+      .split('/')
+      .slice(1);
+    // Must be document-level, so even-numbered elements
+    if (pathArray.length % 2) {
+      throw new Error('The path array must be document-level');
+    }
+
+    // The parent entry is the id of the document
+    const docId = pathArray.pop();
+    // Find the parent of docId
+    let parent = this.database;
+    // Is the parent a collection (if yes, it's an array of documents)
+    let parentIsCollection = false;
+    // Run through the path, creating missing entries
+    for (const entry of pathArray) {
+      if (parentIsCollection) {
+        // find doc in our parent collection (array), or add a new entry
+        // Because we know this isn't the final entry, we immediately
+        // look into _collections for the next entry
+        parent = (
+          parent.find(doc => doc.id === entry) ||
+          // if not found, push a new entry to the collection and return it
+          parent[
+            parent.push({
+              id: entry,
+              _collections: {},
+            }) - 1
+          ]
+        )._collections;
+      } else {
+        // If parent exists, use it else create a new entry and use that
+        parent = parent[entry] || (parent[entry] = []);
+      }
+      parentIsCollection = !parentIsCollection;
+    }
```

Comment:
It took me quite a few read-throughs to understand what is happening here. I think a part of that is there is a lot of logic within logic (i.e. ternaries within if's and or's).  `for` loops in general aren't very descriptive as to what they are doing, so it takes a lot of 🧠 to grok something like this.  I played around with the branch locally, and this logic should be identical but I (personally) find it a bit easier to digest:

```
// The parent entry is the id of the document
const docId = pathArray.pop();
// Find the parent of docId
let parent = this.database;
// Run through the path, creating missing entries

pathArray.forEach((entry, index) => {
  const isCollection = index % 2 === 0;
  if (isCollection) {
    return (parent = parent[entry] || (parent[entry] = []));
  } else {
    const existingDoc = parent.find(doc => doc.id === entry);
    if (existingDoc) {
      return (parent = existingDoc._collections);
    }

    const collections = {};
    parent.push({ id: entry, _collections: collections });
    return (parent = collections);
  }
});
```

Reply:
it was hard to write - it's supposed to be hard to read!!!

This is actually pretty embarrassing.  All I can say is that the fn grew over time in what was probably small, easily justifiable changes.

I had to re-introduce one of those small incremental jumps though.  (and re-wrote as reduce partly to avoid mutating `parent` and partly to sooth my ego)

Reply:
all tests pass (including in my repo)